### ISSUE TITLE
insertAll using mongodb bulk API

### ISSUE THREAD

**Issue by aheckmann** [Automattic/mongoose#723]
batch inserts would be good for perf (with in the model.create method).

obstacles in 2.x:

1) dependence on triggering each inserted documents 'save' hooks. the hook system isn't flexible enough for us to trigger them without actually calling doc.save.
2) doc.save() is tightly coupled to its persistence method to the db (e.g. collection.insert, collection.update). we don't actually want to .insert() it.

we want to trigger hooks without calling doc.save(), then use insertAll.

**Comment by joeltello** [comment]
I am using the create function and I can't find any documentation about error handling. I want to save a collection of documents and it is important for me to keep the documents in the database ONLY IF all the documents get saved. What happens if one of the documents doesn't get saved? or in other words What happens if I get an error in the middle of the insertion?

**Comment by vkarpov15** [comment]
@joeltello `.create()` does `async.parallel()` on documents right now. So if there's an error, you get the first error back, but no information on which documents were inserted. We may improve on this in the future but it's not really a priority right now. PRs welcome :)

**Comment by alsotang** [comment]
it would be a useful improvement.

from http://docs.mongodb.org/manual/core/bulk-write-operations/

> New Bulk methods provide the ability to perform bulk insert, update, and remove operations. MongoDB also supports bulk insert through passing an array of documents to the db.collection.insert() method.

if want to do batch insert, `.insert(docs)` is also a good chioce at this moment.

**Comment by chuks** [comment]
Any progress on this? It would be useful to have a mongoose bulkInsert([array of objecs]) of some sort that does not use Model.create(), and which preserves the schema.

**Comment by vkarpov15** [comment]
Not quite atm. Is there a reason why `Model.create()` doesn't work for you?

**Comment by chuks** [comment]
It is slow for thousands/ millions of inserts. I’d like to use a bulk insert for this

**Comment by vkarpov15** [comment]
Ok we'll make it a priority for 4.3. It's not terribly difficult to do, just haven't had a chance to yet.

**Comment by chuks** [comment]
Hello, why is this closed? Has this bulkInsert feature been implemented yet, or is this no longer a priority?

**Comment by chuks** [comment]
Hello, why is this closed? Has this bulkInsert feature been implemented yet, or is this no longer a priority? Also I see the insertMany feature. Does that do a bulk insert, and is there a limit on the number of documents that can be inserted (bulk) using insertMany?

**Comment by vkarpov15** [comment]
It will be released with 4.4 on monday, `insertMany()` is the new way to do bulk writes with mongodb drivers. I believe the limit is [1000 docs](https://docs.mongodb.org/manual/reference/method/db.collection.insertMany/) right now

**Comment by malcommac** [comment]
Any news about 4.4?

**Comment by chuks** [comment]
I appreciate the information, but was really expecting the no-limit bulk insert. Would this work if, fro example, I used an async module to break it up into groups of 1000 docs, and run in parallel? For example, say I have > 1000 docs I'd like to insert, and I used async to group into arrays of 1000docs, and run them all in parallel?

**Comment by vkarpov15** [comment]
4.4 will be released once I'm done zeroing out issues and making some final logistics checks. You can check release schedules here: https://github.com/Automattic/mongoose/milestones .

@chuks it will work if you break it up in parallel, just tricky for the mongodb server to accept arbitrarily large individual operations.

**Comment by chuks** [comment]
Ok, thanks. One more question --does insertMany() actually do a "bulk insert", i.e. not a loop that goes through sequentially? Right now, I am having success with bulk inserts with thousands of docs with the mongodb driver using:

`Model.collection.insert(modifiedObjectsArray, callback);` 

however, I lose my mongoose model schema, hence I add the additional schema fields to the bulk insert objects (modifiedObjectsArray), before calling it. If insertMany() could replicate exactly what Model.collection.insert()  does, that would be super, instead of worrying about the 1000 docs insertion limit - Model.collection.insert() does not have an insertion limit that I have observed.

Thanks!

**Comment by vkarpov15** [comment]
So `insertMany` uses `Model.collection.insertMany()` under the hood, which is what `Model.collection.insert()` does too. The driver might handle parallelizing >= 1000 docs for you, but I honestly haven't tried yet. Here's the difference between how `insertMany()` and `create()` work:

* `create()` creates a bunch of documents and calls `save()` on them in parallel
* `insertMany()` creates a bunch of documents, calls `.validate()` on them in parallel, and then calls `insertMany()` on them all.

The tradeoffs are that `insertMany()` doesn't trigger pre-save hooks, but it should have better performance because it only makes 1 round-trip to the database rather than 1 for each document.

**Comment by DerekDomino** [comment]
Changelog of 4.4.0 introduces `insertAll()`. While the code and your comments are all about `insertMany()`. I guess we are speaking about the same feature and method ;-)

**Comment by DerekDomino** [comment]
It could make sense to support pre/post methods to `insertMany()`.

E.g., `mongoosastic`relies on post-save() to trigger a synchronization with ElasticSearch. Doing bulk insert could be followed by a bulk synchronization of ES:

```js
mySchema.post('insertMany', function(docs) {
   // Bulk synchronize.
   ...
});
```

What's your opinion?

**Comment by malcommac** [comment]
is there a documentation about insertAll() method? Cannot find anything on http://mongoosejs.com/docs/api.html.

**Comment by vkarpov15** [comment]
@DerekDomino womp yeah it was meant to be `insertMany`. @malcommac see [here](http://mongoosejs.com/docs/api.html#model_Model.insertMany) .

Re hooks, that should be doable. Opened #3846 to track

**Comment by malcommac** [comment]
Just looked at this function but it does not work for me; nothing is returned from the method to caller function.
That's my code:

```
MODEL_EVENT.insertMany(eventsList, function(error, docs) {
	console.log("blabla");
});
```
However console.log is never called :-(

**Comment by malcommac** [comment]
@vkarpov15  does insertMany support upsert to replace existing objs?

**Comment by vkarpov15** [comment]
@malcommac nope, it's just an insert. If there's a conflicting `_id` for instance, the insert will fail.

**Comment by bastianwegge** [comment]
Is the `upsert` support planned or would you recommend doing update's instead?

**Comment by vkarpov15** [comment]
We'll support general `bulkWrite()` at some point in the future https://github.com/Automattic/mongoose/issues/3998 . However, not gonna happen for a few months. For now, I'd recommend just using the mongodb driver directly if you need to do bulk upserts.

**Comment by dciccale** [comment]
@vkarpov15 is there any way to run "create()" sequentially?

The problem is that when I call `create()` passing an array of objects, it does go through the `pre` save, but the counter is not in the correct sequence.

```js
OrderSchema.pre('save', function (next) {
  if (!this.isNew) {
    return next();
  }

  CounterModel.findOneAndUpdate({_id: 'orders'}, {$inc: {n: 1}}, {'new': true, upsert: true}, (err, doc) => {
    if (err) {
      next(err);
    } else {
      this.reference = doc.n;
      next();
    }
  });
});
```

**Output**:

```js
[ { reference: '13' },
  { reference: '21' },
  { reference: '16' },
  { reference: '5' },
  { reference: '9' },
  { reference: '19' },
  { reference: '2' },
  { reference: '10' },
  { reference: '11' },
  { reference: '20' },
  { reference: '8' },
  { reference: '18' },
  { reference: '3' },
  { reference: '15' },
  { reference: '7' },
  { reference: '14' },
  { reference: '4' },
  { reference: '17' },
  { reference: '22' },
  { reference: '6' },
  { reference: '1' },
  { reference: '12' } ]
```

is there any way to achieve the correct sequence of numbers?

I've tried passing the `true` argument and calling `done()` after the `next()` callback, but again same kind of result.

**Comment by vkarpov15** [comment]
Not supported atm. Create just runs a bunch of save() calls in parallel, so no guarantee on order. Can you open a separate issue for that?

**Comment by key-er** [comment]
@vkarpov15 from your comment above in regards to the difference of create() and insertMany() as "doesn't trigger pre-save hooks, but it should have better performance because it only makes 1 round-trip to the database rather than 1 for each document.". 


Does this reason why I am unable to make "unique" work with insertMany. As it doesn't operate on per doc, so unable to see if a specified field value already exists or no..
